keps/1397-evictions-in-background/README.md

```diff
@@ -0,0 +1,1122 @@
+  that were requested to be evicted through the upstream eviction API.
+  In case a descheduler gets unexpectedly restarted any pod annotated
+  with `descheduler.alpha.kubernetes.io/eviction-in-progress` will be added
+  to the cache if not already present.
+  The cache will help with sorting and avoiding a duplicated eviction
+  request of pods that were not annotated
+  with `descheduler.alpha.kubernetes.io/eviction-in-progress` quickly enough.
+  Each item in the cache will have a TTL attribute to address cases where an eviction
+  request was intercepted by an external component but the component failed to annotate the pod.
+- Implement a new built-in sorting plugin: Extend the descheduling framework
+  with a new sorting plugin that will prefer pods
+  with `descheduler.alpha.kubernetes.io/eviction-in-progress` annotation or those
+  that are already in the cache of to-be-evicted pods.
+  The plugin will be enabled when the new functionality is enabled (through a feature gate).
+  Each plugin will have option to either take the new sorting into account or not.
+  The sorting plugin can be executed either in pre-filter or pre-eviction phase.
+- Implement a multi-level sorting: Allow plugins to first sort pods based
+  on eviction-in-progress quality and then based on configured sorting plugins.
+  Preferring pods that are already in a process of eviction over other pods
+  is a crucial step for reducing unnecessary disruptions.
+- Internal counters aware of eviction in progress: At the beginning of each
+  descheduling loop all internal counters will take into account pods
+  that are either annotated with `descheduler.alpha.kubernetes.io/eviction-in-progress`
+  or are present in the cache of to-be-evicted pods.
+
+#### v1alphaN or beta
+- Replace the current eviction API with the new
+  [evacuation API](https://github.com/kubernetes/enhancements/pull/4565) by default.
+  Both `descheduler.alpha.kubernetes.io/request-evict-only` and
+  `descheduler.alpha.kubernetes.io/eviction-in-progress` will be deprecated.
+  Existing deployments will need to remove `descheduler.alpha.kubernetes.io/request-evict-only`
+  annotation to utilize the new evacuation API.
+  Otherwise, the eviction API will be used as a fallback.
+- The evacuation API expects the evacuation request to be reconciled during each descheduling cycle.
+  Given each cycle starts from scratch the descheduler needs to first list
+  all evacuation requests and sort the pods accordingly in each plugin.
+  Other components may create an evacuation request as well.
+  The descheduler needs to take this into account.
+  E.g. by creating a snapshot (per cycle, per strategy) to avoid inconsistencies.
+  An evacuation request may exist without a corresponding pod
+  before the evacuation controller garbage collects the request.
+- The descheduler needs to keep a cache of evacution API requests (CRs)
+  and remove itself from finalizers if the request is no longer needed.
```

Comment:
We are thinking of removing the cancellation feature. If we remove it, then no more work is expected of the descheduler.

Does descheduler have a use case for canceling evictions/evacuations?

Reply:
Only when an evacuation takes too long. Nevertheless, that should not be the case as long as the evacuation is properly reconciled. If the case happens a cluster administrator needs to investigate right away why too many evacuations are pending or not progressing fast enough. As a temporary solution the administrator can increase the descheduler limits to mitigate the situation. Also, TTL functionality will be optional and configurable.

Reply:
Mentioned in the proposal